materials_io/dft.py

```diff
@@ -0,0 +1,95 @@
+from typing import Union, Iterable, Tuple
+from materials_io.base import BaseParser
+from dfttopif import files_to_pif
+from operator import itemgetter
+from glob import glob
+import itertools
+import os
+
+
+# List of files that are known to the VASP parser
+_vasp_file_names = ["outcar", "incar", "chgcar", "wavecar", "wavcar",
+                    "oszicar", "ibzcar", "kpoints", "doscar", "poscar",
+                    "contcar", "vasp_run.xml", "xdatcar"]
+
+
+class DFTParser(BaseParser):
+    """Extract data from Density Functional Theory calculation results
+
+    Uses the `dfttopif <https://github.com/CitrineInformatics/pif-dft>`_ parser to extract
+    metadata from each file
+    """
+
+    def __init__(self, quality_report=False):
+        """Initialize a featurizer
+
+        Args:
+            quality_report (bool): Whether to generate a quality report
+        """
+        self.quality_report = quality_report
+
+    def group(self, paths: Union[str, Iterable[str]], context: dict = None):
+        # Check arguments
+        if isinstance(paths, (str,)):
+            paths = [paths]
+
+        # Clean paths
+        paths = [os.path.abspath(os.path.expanduser(f)) for f in paths]
+
+        # Find all of the files, and attempt to group them
+        files = filter(os.path.isfile, paths)
+        for group in self._group_vasp(files):
+            yield group
+
+        # Recurse into directories
+        for path in filter(os.path.isdir, paths):
+            for group in self.group(glob(os.path.join(path, '*'))):
+                yield group
+
+    def _group_vasp(self, files: Iterable[str]) -> Iterable[Tuple[str, ...]]:
+        """Find groupings of files associated with VASP calculations
+
+        Finds files that start with the name "OUTCAR" (not case sensitive) and groups
+        those files together with any file that share the same postfix
+        (e.g., "OUTCAR.1" and "INCAR.1" are grouped together)
+
+        Args:
+            files ([str]): List of files to be grouped
+        Yields:
+            ((files]): List of VASP files from the same calculation
+        """
+
+        # Get the files that are likely VASP files, which we define as
+        #  those which start with the name of a known vasp file.
+        vasplike_files = []  # List of (path, type, postfix)
+        for file in files:
```

Comment:
I agree that this logic should be in a utility function, as group-certain-filenames is likely to be useful for other parses' grouping.

How does this implementation compare to the MDF pre- and postfix aware grouping logic?
https://github.com/materials-data-facility/connect_server/blob/master/mdf_connect_server/processor/converter.py#L134

Also, we should use a variable name other than `file` (such as `f` or `filename`) to reduce confusion, as `file` was a keyword in Python 2 (and is still syntax-highlighted today).

Reply:
I've gone ahead and added this type of grouping function to a separate model.

I did write this code based your implementation in Connect. However, I stripped off the "parser-matching" functionality (being irrelevant for this case) and adopted a "pick which method works for your need" approach for the grouping utility module rather than a "write a configuration file for the grouping" (I find the "pick which method" approach simpler).

Reply:
I also changed "file" to "filename."

Reply:
I like the `group_by_postfix()` utility, and I agree that the "pick a method" approach is better for this case. At this point what I'm wondering, though, is if it's useful to copy over the Connect logic of handling prefixes as well as postfixes when matching filenames. (This behavior could have a toggle flag to turn off if you think it could cause issues.) The rest of the Connect grouper is likely not useful for the MatIO case, I agree.

Reply:
I'm on the fence about whether to port over the "group by prefix." On one hand, it is nice to get the logic all in one place. But, I'm hesitant to make too many utility operations before we have actual examples of where they are needed.

In this case, i think group by prefix is useful/easy enough to include now. I'll do that, and add tests specific to the grouping methods.